[AAP-48822] Add default LDAP connection options to match documentation

thenets opened the pull request:
## Description

This PR extends the work done in PR #756 by improving how LDAP connection options are handled in the UI while maintaining the default `OPT_REFERRALS` connection option functionality.

https://issues.redhat.com/browse/AAP-48822

**Building on PR #756:**
PR #756 added the default LDAP connection options that our documentation says we have, specifically setting `OPT_REFERRALS` that was previously missing. This PR continues that work by addressing the UI experience.

- What is being changed?
  - Changed the default value of CONNECTION_OPTIONS field from `default_connection_options` to an empty dict `{}`  
  - Updated backend logic to merge the default `OPT_REFERRALS` setting with user-provided options when CONNECTION_OPTIONS is empty or not a dict
  - Builds upon the default `OPT_REFERRALS` connection option established in PR #756

- Why is this change needed?
  - While PR #756 correctly established the default `OPT_REFERRALS` option, users should not be forced to see this default pre-populated in the UI configuration form

- How does this change address the issue?
  - UI now shows empty CONNECTION_OPTIONS field by default (cleaner user experience)
  - Backend properly merges the default `OPT_REFERRALS` setting with any user-provided options
  - Maintains backward compatibility with existing configurations
  - Preserves the `OPT_REFERRALS` default connection option functionality from PR #756

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist
- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [ ] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

## Testing Instructions
### Prerequisites
- LDAP authenticator configuration environment

### Steps to Test
1. Add some debugger or logger to monitor the final value of the `connection_options` variable
2. Run the `test_ldap_config_defaults` ATF test and put a breakpoint before its teardown, this will setup a valid LDAP server into your AAP
3. Check the value of the `connection_options`, which must contain the expected defaults 

### Expected Results
- The `connection_options` must always have the `OPT_REFERRALS` key
- The default values must be used if the user didn't provide a different one

## Additional Context

This PR is a continuation of the LDAP connection options work:
1. **PR #756**: Established the missing default LDAP connection options that match our documentation
2. **This PR (#785)**: Improves the UI experience by not forcing the `OPT_REFERRALS` default to be visible in the form while maintaining all the backend functionality

The combination of both PRs provides the complete solution: proper `OPT_REFERRALS` default that matches documentation + clean UI experience.

### Required Actions
- [ ] Requires documentation updates
- [ ] Requires downstream repository changes  
- [ ] Requires infrastructure/deployment changes
- [ ] Requires coordination with other teams

phess committed:
Disable LDAP referrals by default

john-westcott-iv committed:
Changing default LDAP connection optinos to match documentation

thenets committed:
Do not force default LDAP connection options to the UI

Changed the default value from predefined connection options to an empty
dict, allowing the backend to merge defaults with user-provided options
without forcing defaults in the UI configuration form.

thenets committed:
Fix LDAP config defaults test to match new behavior

Update test_ldap_config_defaults to verify that:
- CONNECTION_OPTIONS field default is empty dict for clean UI
- LDAPSettings properly merges defaults when CONNECTION_OPTIONS is empty
- Default connection options (OPT_REFERRALS, OPT_NETWORK_TIMEOUT) are applied at runtime

This addresses the test failure caused by the change in PR #785 where
CONNECTION_OPTIONS field default was changed back to {} for UI purposes
while still applying defaults in the LDAPSettings class.

thenets committed:
Fix linter

jay-steurer commented:
I've tested this, and I believe this is a better solution given the user point of view.

thenets committed:
Fix the test_ldap_config_defaults test

thenets committed:
Add test for LDAP CONNECTION_OPTIONS user override behavior

This test verifies that user-provided CONNECTION_OPTIONS values
properly override default connection options while preserving
non-overridden defaults and handling edge cases.

The test covers:
- User values overriding defaults (OPT_REFERRALS, OPT_NETWORK_TIMEOUT)
- Additional user options not in defaults
- Mixed scenarios with overrides and defaults
- Invalid CONNECTION_OPTIONS handling

thenets committed:
Make dynamic retrieval of the default values

thenets committed:
no op

john-westcott-iv commented:
In regards to showing the defaults or not, do we know what controller 2.4 did?

thenets commented:
@john-westcott-iv requested this to be reworked.

It will not behave like a real-time default anymore. The default will only appear during the resource creation.

thenets committed:
Remove `OPT_NETWORK_TIMEOUT` related changes

thenets commented:
from the yesterday's discussion:
- we decided to keep this behavior but only for the `OPT_REFERRALS=0`
- the UX improvements must have a dedicated story
- for future work, a migration must add this value if no other was specified into the authenticator settings

bhavenst commented on ansible_base/authentication/authenticator_plugins/ldap.py:
```diff
@@ -434,7 +435,13 @@ def __init__(self, prefix: str = 'AUTH_LDAP_', defaults: dict = {}):
         # Connection options need to be set as {"integer": "value"} but our configuration has {"friendly_name": "value"} so we need to convert them
-        connection_options = defaults.get('CONNECTION_OPTIONS', {})
+        connection_options = defaults.get('CONNECTION_OPTIONS')
+        if not isinstance(connection_options, dict):
```
Seems like this change is just to dump a non-dict if it shows up here?  Maybe a warning in the log if that happens?

thenets replied:
Just added it.

thenets committed:
Add warning for invalid CONNECTION_OPTIONS in LDAPSettings

This change introduces a warning log when the CONNECTION_OPTIONS is not a dictionary, helping to identify potential misconfigurations in LDAP settings.
